feat(clean): add clean_df function

AndyWangSFU opened the pull request:
# Description

`clean_df` function: conduct a set of operations that would be useful for cleaning and standardizing a full Pandas DataFrame. Closes #503.

# How Has This Been Tested?

I have tested this function using a few real-world datasets. I will also add my test function later.


# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have already squashed the commits and make the commit message conform to the project standard.
- [x] I have already marked the commit with "BREAKING CHANGE" or "Fixes #" if needed.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

yxie66 commented on dataprep/clean/clean_df.py:
```diff
@@ -0,0 +1,414 @@
+        The desired way to check data types.
+        * If ’semantic’, then perform a column-wise semantic and atomic type detection.
+        * If 'atomic', then return the best inferred atomic data type from Python default ones.
+        * If 'none', then no results will be returned.
```
I think the input can't be 'none' here

AndyWangSFU replied:
I guess "none" means that the users do not want to perform any data type detection?

yxie66 replied:
> I guess "none" means that the users do not want to perform any data type detection?

Based on my understanding of the implementation here, the input parameter of `_infer_data_type_df()` will not be 'none'. If it's not 'semantic', you will perform the atomic detection. Which means it can only be 'semantic' and 'atomic'?

qidanrui replied:
> > I guess "none" means that the users do not want to perform any data type detection?
> 
> Based on my understanding of the implementation here, the input parameter of `_infer_data_type_df()` will not be 'none'. If it's not 'semantic', you will perform the atomic detection. Which means it can only be 'semantic' and 'atomic'?

Yeah, I agree with Yi's opinion. It seems that the `_infer_data_type_df()` function doesn't have an option for the parameter 'none'. 
<img width="798" alt="Screen Shot 2021-04-20 at 4 32 15 PM" src="https://user-images.githubusercontent.com/10005367/115364433-f9816b80-a1f5-11eb-9c23-a3f71bd12b11.png">
How about changing the code like this?
<img width="797" alt="Screen Shot 2021-04-20 at 4 36 17 PM" src="https://user-images.githubusercontent.com/10005367/115364997-8a584700-a1f6-11eb-95f5-350a7abdf82e.png">

AndyWangSFU replied:
Thanks Danrui! Greatly appreciated. I did not find this typo before.

yxie66 commented on dataprep/clean/clean_df.py:
```diff
@@ -0,0 +1,414 @@
+            print(f"\tMemory reducted from {old_stat} to {nclnd}. New size: ({pclnd}%)")
+        else:
+            print("Downcast Memory not Performed.")
+
```
Maybe it's better to add an illegal checking here, to avoid illegal call

AndyWangSFU replied:
Thanks Yi. Sorry I did not get it, what type of illegal call it can possibly be?

yxie66 replied:
What I mean is that input `option` parameter can be an illegal string.

AndyWangSFU replied:
I am not very sure whether it could happen, as the `_create_report()` function is only callable inside `clean_df()` by the `if report: (line 131)` condition. But users might directly call this function accidentally and receive unexpected errors. So I am OK to add a check condition inside this function.

By the way, it just reminds me that users could get confused by this `_create_report()` with the other report functions in `utils.py`? Is it necessary to rename this function to `_create_report_clean_df()`?

yxie66 replied:
Yes, what I'm referring to is potential accidental touch by users. But I'm OK with either way so just let me know your decision. For the function name, do you think `create_report_df()` or `create_df_report()` will work?

AndyWangSFU replied:
I feel like ` create_report_df()` works. We can also listen to Danrui's opinion!

AndyWangSFU replied:
Fixed in the newest commit!

qidanrui commented:
I think the `pass` in `_standardize_missing_values_df()` is redundant, because your if statements are enough. 
![image](https://user-images.githubusercontent.com/10005367/115366696-2171ce80-a1f8-11eb-8122-cbdd7d4d06c6.png)
If you think there still should be placeholder, how about changing it like this:
<img width="431" alt="Screen Shot 2021-04-20 at 4 50 44 PM" src="https://user-images.githubusercontent.com/10005367/115367189-8deccd80-a1f8-11eb-9812-a8453afa6927.png">
I think it is more concrete~

qidanrui commented:
Sorry for that I'm little confusing why there are two very similar functions returning opposite values. Is there any possibility of combining them?
<img width="675" alt="Screen Shot 2021-04-20 at 4 59 36 PM" src="https://user-images.githubusercontent.com/10005367/115368584-cc36bc80-a1f9-11eb-9c5c-750e5c4e1a7a.png">

qidanrui commented:
I changed some return types according to Andy's logic:
* Changing the return type of `clean_df()` function from `pd.DataFrame` to `Union[Tuple[pd.DataFrame, pd.DataFrame], pd.DataFrame]`. That is because if the parameter`data_type_detection = 'none'`, the function returns only one dataframe (origin dataframe). However, if the parameter`data_type_detection != 'none'`, the function returns two dataframes. Thus I think it is better to use the union returning type.
<img width="501" alt="Screen Shot 2021-04-20 at 5 20 56 PM" src="https://user-images.githubusercontent.com/10005367/115371682-c68ea600-a1fc-11eb-9be6-8f24291f4686.png">
* Changing the return type of `_infer_semantic_data_type()` from `Any` to `str`
<img width="494" alt="Screen Shot 2021-04-20 at 5 36 29 PM" src="https://user-images.githubusercontent.com/10005367/115373964-f474ea00-a1fe-11eb-88e6-302862ff74e2.png">
* Changing the return type of `_infer_atomic_data_type()` from `Any` to `str`
<img width="471" alt="Screen Shot 2021-04-20 at 5 36 45 PM" src="https://user-images.githubusercontent.com/10005367/115373999-fc348e80-a1fe-11eb-8771-b5df0cd275bc.png">

No more comments for this function @yxie66. I think after Andy's small refining we can merge it.

AndyWangSFU commented:
> I think the `pass` in `_standardize_missing_values_df()` is redundant, because your if statements are enough.
> ![image](https://user-images.githubusercontent.com/10005367/115366696-2171ce80-a1f8-11eb-8122-cbdd7d4d06c6.png)
> If you think there still should be placeholder, how about changing it like this:
> <img alt="Screen Shot 2021-04-20 at 4 50 44 PM" width="431" src="https://user-images.githubusercontent.com/10005367/115367189-8deccd80-a1f8-11eb-9812-a8453afa6927.png">
> I think it is more concrete~

This makes sense! I fixed it in the latest commit, using the second suggestion.

AndyWangSFU commented:
> I changed some return types according to Andy's logic:
> 
> * Changing the return type of `clean_df()` function from `pd.DataFrame` to `Union[Tuple[pd.DataFrame, pd.DataFrame], pd.DataFrame]`. That is because if the parameter`data_type_detection = 'none'`, the function returns only one dataframe (origin dataframe). However, if the parameter`data_type_detection != 'none'`, the function returns two dataframes. Thus I think it is better to use the union returning type.
> 
> <img alt="Screen Shot 2021-04-20 at 5 20 56 PM" width="501" src="https://user-images.githubusercontent.com/10005367/115371682-c68ea600-a1fc-11eb-9be6-8f24291f4686.png">
> 
> * Changing the return type of `_infer_semantic_data_type()` from `Any` to `str`
> <img alt="Screen Shot 2021-04-20 at 5 36 29 PM" width="494" src="https://user-images.githubusercontent.com/10005367/115373964-f474ea00-a1fe-11eb-88e6-302862ff74e2.png">
> 
> * Changing the return type of `_infer_atomic_data_type()` from `Any` to `str`
> <img alt="Screen Shot 2021-04-20 at 5 36 45 PM" width="471" src="https://user-images.githubusercontent.com/10005367/115373999-fc348e80-a1fe-11eb-8771-b5df0cd275bc.png">
> 
> No more comments for this function @yxie66. I think after Andy's small refining we can merge it.

Point 1 totally makes sense and I modified it. Regarding the return types, telling from the pylint test, I guess the `infer_type()` function from Pandas does not always return `str` and the default is `Any`. It might be better to keep this.

AndyWangSFU commented:
> Sorry for that I'm little confusing why there are two very similar functions returning opposite values. Is there any possibility of combining them?
> <img alt="Screen Shot 2021-04-20 at 4 59 36 PM" width="675" src="https://user-images.githubusercontent.com/10005367/115368584-cc36bc80-a1f9-11eb-9c5c-750e5c4e1a7a.png">

This is a very good question. I was hoping to combine them together, and there is definitely a way to achieve it by adding one more parameter. For example, I tried `check_values(data: str, valid: bool)` to switch the check conditions between valid or invalid.

However, the code efficiency drops a lot if adding this check condition, as it is mapped to every string and every column. Also, I was thinking that it would be more clear to leave two useful functions, because users can directly call one of them to use. There are equal scenarios when checking "valid" or "null". It will be more confusing to add optional parameters.

Based on these two reasons, I finally decided to keep both of them. Hope it makes sense!

qidanrui commented:
> > Sorry for that I'm little confusing why there are two very similar functions returning opposite values. Is there any possibility of combining them?
> > <img alt="Screen Shot 2021-04-20 at 4 59 36 PM" width="675" src="https://user-images.githubusercontent.com/10005367/115368584-cc36bc80-a1f9-11eb-9c5c-750e5c4e1a7a.png">
> 
> This is a very good question. I was hoping to combine them together, and there is definitely a way to achieve it by adding one more parameter. For example, I tried `check_values(data: str, valid: bool)` to switch the check conditions between valid or invalid.
> 
> However, the code efficiency drops a lot if adding this check condition, as it is mapped to every string and every column. Also, I was thinking that it would be more clear to leave two useful functions, because users can directly call one of them to use. There are equal scenarios when checking "valid" or "null". It will be more confusing to add optional parameters.
> 
> Based on these two reasons, I finally decided to keep both of them. Hope it makes sense!

Thanks for your replying Andy! Now I think they're reasonable~

qidanrui commented:
> > I changed some return types according to Andy's logic:
> > 
> > * Changing the return type of `clean_df()` function from `pd.DataFrame` to `Union[Tuple[pd.DataFrame, pd.DataFrame], pd.DataFrame]`. That is because if the parameter`data_type_detection = 'none'`, the function returns only one dataframe (origin dataframe). However, if the parameter`data_type_detection != 'none'`, the function returns two dataframes. Thus I think it is better to use the union returning type.
> > 
> > <img alt="Screen Shot 2021-04-20 at 5 20 56 PM" width="501" src="https://user-images.githubusercontent.com/10005367/115371682-c68ea600-a1fc-11eb-9be6-8f24291f4686.png">
> > 
> > * Changing the return type of `_infer_semantic_data_type()` from `Any` to `str`
> > 
> > <img alt="Screen Shot 2021-04-20 at 5 36 29 PM" width="494" src="https://user-images.githubusercontent.com/10005367/115373964-f474ea00-a1fe-11eb-88e6-302862ff74e2.png">
> > 
> > * Changing the return type of `_infer_atomic_data_type()` from `Any` to `str`
> > 
> > <img alt="Screen Shot 2021-04-20 at 5 36 45 PM" width="471" src="https://user-images.githubusercontent.com/10005367/115373999-fc348e80-a1fe-11eb-8771-b5df0cd275bc.png">
> > No more comments for this function @yxie66. I think after Andy's small refining we can merge it.
> 
> Point 1 totally makes sense and I modified it. Regarding the return types, telling from the pylint test, I guess the `infer_type()` function from Pandas does not always return `str` and the default is `Any`. It might be better to keep this.

Got! Reasonable to me~Thanks Andy!
